sdk/iot/core/doc/mqtt_state_machine.md

```diff
@@ -136,3 +138,120 @@ _Example:_
+{
+    // success case 
+}
+else
+{
+    if (az_iot_provisioning_client_is_retriable_status(status))
+    {
+        // retry
+    }
+    else
+    {
+        // fail
+    }
+}
+```
+
+#### Retry timing
+
+Network timeouts and the MQTT keep-alive interval should be configured considering tradeoffs between how fast network issues are detected vs traffic overheads. [This document](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-mqtt-support#default-keep-alive-timeout) describes the recommended keep-alive timeouts as well as the minimum idle timeout supported by Azure IoT services.
+
+For connectivity issues at all layers (TCP, TLS, MQTT) as well as cases where there is no `retry-after` sent by the service, we suggest using an exponential back-off with random jitter function. `_az_retry_calc_delay` is available in Azure SDK Core:
+
+```C
+// random_ms is a random value between 0 and max_jitter_msec
+// The previous operation took operation_msec
+
+int32_t total_delay_msec =  random_msec + _az_retry_calc_delay(attempt, min_retry_delay_msec, max_retry_delay_msec);
+int32_t delay_msec = max(operation_msec - total_delay_msec, 0); 
+```
+
+_Note 1_: The network stack may have used more time than the recommended delay before timing out. (e.g. The operation timed out after 2 minutes while the delay between operations is 1 second). In this case there is no need to delay the next operation.
+
+_Note 2_: To determine the parameters of the exponential with back-off retry strategy, we recommend modeling the network characteristics (including failure-modes). Compare the results with defined SLAs for device connectivity (e.g. 1M devices must be connected in under 30 minutes) and with the available IoT Azure scale (especially consider _throttling_, _quotas_ and maximum _requests/connects per second_).
+
+In the absence of modeling, we recommend the following default:
+```C
+    min_retry_delay_msec =   1000;
+    max_retry_delay_msec = 100000;
+    max_jitter_msec      =   5000;
+```
+
+For service-level errors, the Provisioning Service is providing a `retry-after` (in seconds) parameter:
+
+```C
+// az_iot_provisioning_client_received_topic_payload_parse was successful and created a az_iot_provisioning_client_register_response response
+
+int32_t delay_ms;
+if ( response.retry_after_seconds > 0 )
+{
+    delay_ms = response.retry_after_seconds;
+}
+else
+{
+    delay_ms = random_ms + _az_retry_calc_delay(attempt, min_retry_delay_msec, max_retry_delay_msec);
+}
+```
+
+#### Suggested Retry Strategy
+
+Combining the functions above we recommend the following flow:
+
+![image](resources/iot_retry_flow.png "IoT MQTT Retry Flow")
+
+When devices are using IoT Hub without Provisioning Service, we recommend attempting to rotate the IoT Credentials (SAS Token or X509 Certificate) on authentication issues.
+
+_Note:_ Authentication issues observed in the following cases do not require credentials to be rotated and require further user actions:

```

Comment:
>user [](start = 123, length = 4)

Nit: I think we should avoid word "user" here since it's unclear if this is end user or the "user of our SDK, aka dev".